File: ios/Packages/RideLinkCore/Sources/RideLinkCore/Transfer/TransferModels.swift
import Foundation

/// ADR-023 / PROTOCOL §8.2 — one transfer's lifecycle. `.complete`, `.failed` and `.cancelled` are
/// terminal: `TransferReducer` never transitions out of them again (brief §43).
///
/// Raw values match the Kotlin enum constant names exactly, since `protocol/vectors/transfer-fsm/`
/// names them that way and both platforms read the same file.
public enum TransferStatus: String, Sendable, Equatable, CaseIterable {
    case idle = "IDLE"
    case queued = "QUEUED"
    case negotiating = "NEGOTIATING"
    case transferring = "TRANSFERRING"
    case verifying = "VERIFYING"
    case complete = "COMPLETE"
    case failed = "FAILED"
    case cancelled = "CANCELLED"

    public var isTerminal: Bool { self == .complete || self == .failed || self == .cancelled }
}

/// Machine-readable transfer failures — never a human string parsed to decide behaviour.
public enum TransferError: String, Sendable, Equatable, CaseIterable {
    case notFound = "NOT_FOUND"
    case notAuthorized = "NOT_AUTHORIZED"
    case invalidRequest = "INVALID_REQUEST"
    case unsupported = "UNSUPPORTED"
    case fileChanged = "FILE_CHANGED"
    case sizeMismatch = "SIZE_MISMATCH"
    case hashMismatch = "HASH_MISMATCH"
    case diskFull = "DISK_FULL"
    case ioError = "IO_ERROR"
    case connectionLost = "CONNECTION_LOST"
    case cancelled = "CANCELLED"
    case protocolError = "PROTOCOL_ERROR"
}

/// One input to `TransferReducer`.
public enum TransferEvent: Sendable, Equatable {
    case enqueued
    case dequeued
    case offerReceived(sizeBytes: Int64, chunkCount: Int)
    case offerRejected(error: TransferError)
    case bytesReceived(bytes: Int64)
    case sizeMismatchDetected
    case allBytesReceived
    case hashVerified(matches: Bool)
    case ioErrorDetected
    case diskFullDetected
    /// The bulk connection or the control session it is bound to (ADR-023) dropped.
    case connectionLost
    /// ADR-023 §3 — the control session/generation that authorised this transfer no longer matches.
    case sessionInvalidated
    case cancelled
}

/// One side-effect `TransferReducer` asks the caller to perform. A diff, never a restatement.
public enum TransferAction: Sendable, Equatable {
    case sendTransferRequest
    case openBulkConnection
    case writeChunkToPart
    case reportProgress(bytes: Int64)
    case computeHashFromDisk
    /// The only path to a verified cache entry (ADR-023 §6) — atomic `.part` -> final rename.
    case promoteCacheEntry
    case deletePartFile
    case closeBulkConnection
    case notifyUi(status: TransferStatus, error: TransferError? = nil)
}

public struct TransferTransition: Sendable, Equatable {
    public let status: TransferStatus
    public let actions: [TransferAction]
    public let error: TransferError?

    public init(status: TransferStatus, actions: [TransferAction], error: TransferError?) {
        self.status = status
        self.actions = actions
        self.error = error
    }
}

File: android/core/src/main/kotlin/com/ridelink/core/transfer/TransferModels.kt
package com.ridelink.core.transfer

/**
 * ADR-023 / PROTOCOL §8.2 — one transfer's lifecycle. `COMPLETE`, `FAILED` and `CANCELLED` are
 * terminal: [TransferReducer] never transitions out of them again (brief §43).
 */
enum class TransferStatus {
    IDLE,
    QUEUED,
    NEGOTIATING,
    TRANSFERRING,
    VERIFYING,
    COMPLETE,
    FAILED,
    CANCELLED,
    ;

    val isTerminal: Boolean get() = this == COMPLETE || this == FAILED || this == CANCELLED
}

/** Machine-readable transfer failures — never a human string parsed to decide behaviour. */
enum class TransferError {
    NOT_FOUND,
    NOT_AUTHORIZED,
    INVALID_REQUEST,
    UNSUPPORTED,
    FILE_CHANGED,
    SIZE_MISMATCH,
    HASH_MISMATCH,
    DISK_FULL,
    IO_ERROR,
    CONNECTION_LOST,
    CANCELLED,
    PROTOCOL_ERROR,
}

/** One input to [TransferReducer]. */
sealed class TransferEvent {
    object Enqueued : TransferEvent()

    object Dequeued : TransferEvent()

    data class OfferReceived(
        val sizeBytes: Long,
        val chunkCount: Int,
    ) : TransferEvent()

    data class OfferRejected(
        val error: TransferError,
    ) : TransferEvent()

    data class BytesReceived(
        val bytes: Long,
    ) : TransferEvent()

    object SizeMismatchDetected : TransferEvent()

    object AllBytesReceived : TransferEvent()

    data class HashVerified(
        val matches: Boolean,
    ) : TransferEvent()

    object IoErrorDetected : TransferEvent()

    object DiskFullDetected : TransferEvent()

    /** The bulk connection or the control session it is bound to (ADR-023) dropped. */
    object ConnectionLost : TransferEvent()

    /** ADR-023 §3 — the control session/generation that authorised this transfer no longer matches. */
    object SessionInvalidated : TransferEvent()

    object Cancelled : TransferEvent()
}

/** One side-effect [TransferReducer] asks the caller to perform. A diff, never a restatement. */
sealed class TransferAction {
    object SendTransferRequest : TransferAction()

    object OpenBulkConnection : TransferAction()

    object WriteChunkToPart : TransferAction()

    data class ReportProgress(
        val bytes: Long,
    ) : TransferAction()

    object ComputeHashFromDisk : TransferAction()

    /** The only path to a verified cache entry (ADR-023 §6) — atomic `.part` -> final rename. */
    object PromoteCacheEntry : TransferAction()

    object DeletePartFile : TransferAction()

    object CloseBulkConnection : TransferAction()

    data class NotifyUi(
        val status: TransferStatus,
        val error: TransferError? = null,
    ) : TransferAction()
}

data class TransferTransition(
    val status: TransferStatus,
    val actions: List<TransferAction>,
    val error: TransferError?,
)
